Add the ability to read user provided debug files

ghadi-rahme opened the pull request:
This PR adds the ability to the user to provide their own ddeb or vmlinux file. The `--debug-file` parameter was added to specify the path of the file.

Not sure if the unit test is necessary but I added it just in case.

Updated the README as well to display the new bash output.

Added a `TODO` to allow the user to specify `--no-debuginfod` and  `--no-pullpkg` while `--debug-file` is specified, although this does not affect how the program runs if they are specified or not when `--debug-file` is specified.

For detecting the file type, I though about adding an extra check for vmlinux files in `debug_file_type()` where it would check if the file start with `vmlinux` as well instead of simply defaulting to it if the file is not a ddeb
 
Closes: SET-941

mustafakemalgilor commented:
Hi @ghadi-rahme , thanks for the patch!

> Added a `TODO` to allow the user to specify `--no-debuginfod` and `--no-pullpkg` while `--debug-file` is specified, although this does not affect how the program runs if they are specified or not when `--debug-file` is specified.

I think both "--no-debuginfod" and "--no-pullpk" should be implicitly `true` when `--debug-file` is specified, because we don't want hotkdump to fetch the debug symbol automatically, we want it to use the user-specified one.

> For detecting the file type, I though about adding an extra check for vmlinux files in `debug_file_type()` where it would check if the file start with `vmlinux` as well instead of simply defaulting to it if the file is not a ddeb

I agree, having a check in place would be nice.

ghadi-rahme commented:
Thank you for the feedback! 
I'll start working on those and fix the pylint errors.

mustafakemalgilor commented on hotkdump/core/hotkdump.py:
```diff
@@ -493,17 +503,22 @@ def run(self):
                     print(f"{key}={self.kdump_file.vmcoreinfo.get(key)}")
                 return
 
-            extracted_vmlinux = None
+            dbg_type = self.debug_file_type()
+            vmlinux_ddeb = self.params.debug_file if dbg_type == "ddeb" else None
+            extracted_vmlinux = self.params.debug_file if dbg_type == "vmlinux" else None
 
-            if not self.params.no_debuginfod:
-                extracted_vmlinux = self.maybe_download_vmlinux_via_debuginfod()
+            # Avoid downloading debug symbols if a debug file was specified.
+            if not dbg_type:
+                if not self.params.no_debuginfod:
+                    extracted_vmlinux = self.maybe_download_vmlinux_via_debuginfod()
 
-            if not extracted_vmlinux and not self.params.no_pullpkg:
-                vmlinux_ddeb = self.maybe_download_vmlinux_via_pullpkg()
-                if vmlinux_ddeb == "":
-                    logging.error("vmlinux ddeb dowload failed.")
-                    return
+                if not extracted_vmlinux and not self.params.no_pullpkg:
+                    vmlinux_ddeb = self.maybe_download_vmlinux_via_pullpkg()
+                    if vmlinux_ddeb == "":
+                        logging.error("vmlinux ddeb dowload failed.")
+                        return
```
This part can be much simpler if we do the following:

- make a function named `maybe_get_user_specified_vmlinux` and return the vmlinux (automatically extract if .ddeb)
- change `maybe_download_vmlinux_via_pullpkg` to return a path to extracted vmlinux (move extract_vmlinux_ddeb call to maybe_download_vmlinux_via_pullpkg)
- move self.params.no_debuginfod and self.params.no_pullpkg checks to respective maybe_download_vmlinux*  functions and return None when True

... then we can simply do the following:

```python
extracted_vmlinux = self.maybe_get_user_specified_vmlinux() or self.maybe_download_vmlinux_via_debuginfod() or self.maybe_download_vmlinux_via_pullpkg()
```

mustafakemalgilor commented on hotkdump/main.py:
```diff
@@ -70,6 +70,12 @@ def main():
+    ap.add_argument(
+        "--debug-file",
+        required=False,
+        help="Specify the debug file to use. Only ddebs and vmlinux files are supported.",
```
We should also mention here it implies no_debuginfod and no_pullpkg.

mustafakemalgilor commented on hotkdump/core/hotkdump.py:
```diff
@@ -59,13 +59,14 @@ class HotkdumpParameters:
-        if all([self.no_debuginfod, self.no_pullpkg]):
+        if all([self.no_debuginfod, self.no_pullpkg, not self.debug_file]):
             raise ExceptionWithLog(
                 "At least one download method must be enabled (debuginfod, pullpkg)!"
```
... or a file must be specified

mustafakemalgilor commented on hotkdump/core/hotkdump.py:
```diff
@@ -357,42 +358,43 @@ def maybe_download_vmlinux_via_debuginfod(self):
-                return line
-
-            logging.info("debuginfod-find: download for BUILD-ID `%s` failed with `%s`", build_id, line)
+            if not self.params.no_debuginfod:
```
Same for the other maybe_ methods.

mustafakemalgilor commented:
Thanks, @ghadi-rahme! I'll rebase and merge this patch.

mustafakemalgilor commented:
> Thanks, @ghadi-rahme! I'll rebase and merge this patch.

@ghadi-rahme rebase seems to be trivial, can you do it?

ghadi-rahme commented:
@mustafakemalgilor yes I have already rebased it locally, however while reviewing my patch before resending it I noticed a bug. I have fixed it and will also need to change a unit test as well. 
The bug is returning the ddeb path rather than the vmlinux path in `maybe_download_vmlinux_via_pullpkg(self):`

I also ran into some issues running tox where it was complaining that the jinja2 module was not installed... deleting and reinstalling jinja2 fixed it
